src/dbnode/encoding/m3tsz/iterator.go

```diff
@@ -384,30 +225,365 @@ func (it *readerIterator) hasNext() bool {
+	if !it.SkipMarkers {
+		dod, success, err := it.tryReadMarker(stream)
+		if err != nil {
+			return 0, err
+		}
+		if it.Done {
+			return 0, nil
+		}
+
+		if success {
+			return dod, nil
+		}
+	}
+
+	tes, exists := it.Opts.TimeEncodingSchemes()[it.TimeUnit]
+	if !exists {
+		return 0, fmt.Errorf("time encoding scheme for time unit %v doesn't exist", it.TimeUnit)
+	}
+
+	return it.readDeltaOfDelta(stream, tes)
+}
+
+func (it *TimestampIterator) readDeltaOfDelta(
+	stream encoding.IStream, tes encoding.TimeEncodingScheme) (time.Duration, error) {
+	if it.TimeUnitChanged {
+		// NB(xichen): if the time unit has changed, always read 64 bits as normalized
+		// dod in nanoseconds.
+		dodBits, err := stream.ReadBits(64)
+		if err != nil {
+			return 0, err
+		}
+
+		dod := encoding.SignExtend(dodBits, 64)
+		return time.Duration(dod), nil
+	}
+
+	cb, err := stream.ReadBits(1)
+	if err != nil {
+		return 0, err
+	}
+	if cb == tes.ZeroBucket().Opcode() {
+		return 0, nil
+	}
+
+	buckets := tes.Buckets()
+	for i := 0; i < len(buckets); i++ {
+		nextCB, err := stream.ReadBits(1)
+		if err != nil {
+			return 0, nil
+		}
+
+		cb = (cb << 1) | nextCB
+		if cb == buckets[i].Opcode() {
+			dodBits, err := stream.ReadBits(buckets[i].NumValueBits())
+			if err != nil {
+				return 0, err
+			}
+
+			dod := encoding.SignExtend(dodBits, buckets[i].NumValueBits())
+			timeUnit, err := it.TimeUnit.Value()
+			if err != nil {
+				return 0, nil
+			}
+
+			return xtime.FromNormalizedDuration(dod, timeUnit), nil
+		}
+	}
+
+	numValueBits := tes.DefaultBucket().NumValueBits()
+	dodBits, err := stream.ReadBits(numValueBits)
+	if err != nil {
+		return 0, err
+	}
+
+	dod := encoding.SignExtend(dodBits, numValueBits)
+	timeUnit, err := it.TimeUnit.Value()
+	if err != nil {
+		return 0, nil
+	}
+
+	return xtime.FromNormalizedDuration(dod, timeUnit), nil
+}
+
+func (it *TimestampIterator) readAnnotation(stream encoding.IStream) error {
+	antLen, err := it.readVarint(stream)
+	if err != nil {
+		return err
+	}
+
+	// NB: we add 1 here to offset the 1 we subtracted during encoding.
+	antLen = antLen + 1
+	if antLen <= 0 {
+		return fmt.Errorf("unexpected annotation length %d", antLen)
+	}
+
+	// TODO(xichen): use pool to allocate the buffer once the pool diff lands.
+	buf := make([]byte, antLen)
+	n, err := stream.Read(buf)
+	if err != nil {
+		return err
+	}
+	if n != antLen {
+		return fmt.Errorf(
+			"expected to read %d annotation bytes but read: %d",
+			antLen, n)
+	}
+	it.PrevAnt = buf
+
+	return nil
+}
+
+func (it *TimestampIterator) readVarint(stream encoding.IStream) (int, error) {
+	res, err := binary.ReadVarint(stream)
+	return int(res), err
+}
+
+func (it *TimestampIterator) tryPeekBits(stream encoding.IStream, numBits int) (uint64, bool) {
+	res, err := stream.PeekBits(numBits)
+	if err != nil {
+		return 0, false
+	}
+	return res, true
+}
+
+// ReadXOR reads the next XOR value.
+func ReadXOR(stream encoding.IStream, prevXOR uint64) (uint64, error) {
```

Comment:
Would it not make sense to make this an `XORIterator`  and then you wouldn't have to pass in the `prevXOR`?